src/lib/FreeC/Backend/Coq/Base.hs

```diff
@@ -92,6 +129,89 @@ partialUndefined = Coq.bare "undefined"
 partialError :: Coq.Qualid
 partialError = Coq.bare "error"
 
+-------------------------------------------------------------------------------
+-- Sharing                                                                   --
+-------------------------------------------------------------------------------
+-- | The Coq identifier for the @Share@ module.
+shareModuleIdent :: Coq.Ident
+shareModuleIdent = Coq.ident "Share"
+
+-- | The Coq identifier for the @Injectable@ type class.
+injectable :: Coq.Qualid
+injectable = Coq.bare "Injectable"
+
+-- | The Coq binder for the @Injectable@ type class.
+injectableBinder :: Coq.Binder
+injectableBinder = Coq.Generalized Coq.Implicit
+  $ Coq.app (Coq.Qualid injectable)
+  $ map Coq.Qualid [ Coq.Qualified shareModuleIdent shapeIdent
+                   , Coq.Qualified shareModuleIdent posIdent
+                   , shape
+                   , pos
+                   ]
+
+-- | The Coq identifier for the @Strategy@ type class.
+strategy :: Coq.Qualid
+strategy = Coq.bare "Strategy"
+
+-- | The Coq identifier for the argument of the @Strategy@ type class.
+strategyArg :: Coq.Qualid
+strategyArg = Coq.bare "S"
+
+-- | The Coq binder for the @Strategy@ type class.
+strategyBinder :: Coq.Binder
+strategyBinder = Coq.typedBinder' Coq.Ungeneralizable Coq.Explicit strategyArg
+  $ Coq.app (Coq.Qualid strategy) [Coq.Qualid shape, Coq.Qualid pos]
+
+-- | The Coq binder for the @ShareableArgs@ type class.
+shareableArgs :: Coq.Qualid
+shareableArgs = Coq.bare "ShareableArgs"
+
+-- | The Coq binder for the @ShareableArgs@ type class with the type variable
+--   with the given name.
+shareableArgsBinder :: Coq.Qualid -> Coq.Binder
+shareableArgsBinder typeArg = Coq.Generalized Coq.Implicit
+  $ Coq.app (Coq.Qualid shareableArgs)
+  $ map Coq.Qualid [shape, pos, typeArg]
+
+-- | The Coq identifier for an implicit argument.
+implicitArg :: Coq.Qualid
+implicitArg = Coq.bare $ showP Coq.Underscore
```

Comment:
The idea of my [original suggestion](https://github.com/FreeProving/free-compiler/pull/179#discussion_r484566419) was to use no bare identifier for `_` but the corresponding AST node. Since `implicitArg` is always wrapped with a `Coq.Qualid` constructor, you can just return a `Coq.Term` directly.

```suggestion
-- | The Coq identifier for an implicit argument.
implicitArg :: Coq.Term
implicitArg = Coq.Underscore
```

For example

```haskell
implicitEffectArgs' = map Coq.Qualid $ concatMap Coq.Base.selectImplicitArgs effects
```

would become

```haskell
implicitEffectArgs' = concatMap Coq.Base.selectImplicitArgs effects
```